pallets/automation-price/src/lib.rs

```diff
@@ -0,0 +1,699 @@
+			if let Some(_) = Self::get_task((asset.clone(), task_id.clone())) {
+				Err(Error::<T>::DuplicateTask)?
+			}
+			if let Some(mut asset_tasks) = Self::get_scheduled_tasks((asset.clone(), direction.clone(), trigger_percentage.clone())) {
+				if let Err(_) = asset_tasks.try_push(task_id.clone()) {
+					Err(Error::<T>::MaxTasksReached)?
+				}
+				<ScheduledTasks<T>>::insert((asset, direction, trigger_percentage), asset_tasks);
+			} else {
+				let scheduled_tasks: BoundedVec<T::Hash, T::MaxTasksPerSlot> = vec![task_id.clone()].try_into().unwrap();
+				<ScheduledTasks<T>>::insert((asset, direction, trigger_percentage), scheduled_tasks);
+			}
+			Ok(task_id)
+		}
+
+		/// Validate and schedule task.
+		/// This will also charge the execution fee.
+		pub fn validate_and_schedule_task(
+			who: T::AccountId,
+			provided_id: Vec<u8>,
+			asset: Vec<u8>,
+			direction: u8,
+			trigger_percentage: u128,
+		) -> Result<(), Error<T>> {
+			if provided_id.len() == 0 {
+				Err(Error::<T>::EmptyProvidedId)?
+			}
+			let asset_target_price: u128 = match Self::get_asset_target_price(asset.clone()) {
+				None => Err(Error::<T>::AssetNotSupported)?,
+				Some(asset_price) => asset_price,
+			};
+			let last_asset_price: u128 = match Self::get_asset_price(asset.clone()) {
+				None => Err(Error::<T>::AssetNotSupported)?,
+				Some(asset_price) => asset_price,
+			};
+			if (direction == 0) & (last_asset_price < asset_target_price) {
+				let last_asset_percentage = Self::calculate_asset_percentage(last_asset_price, asset_target_price);
+				if trigger_percentage > last_asset_percentage {
+					Err(Error::<T>::AssetNotInTriggerableRange)?
+				}
+			}
+			if (direction == 1) & (last_asset_price > asset_target_price) {
+				let last_asset_percentage = Self::calculate_asset_percentage(last_asset_price, asset_target_price);
+				if trigger_percentage < last_asset_percentage {
+					Err(Error::<T>::AssetNotInTriggerableRange)?
+				}
+			}
+			let task_id =
+				Self::schedule_task(who.clone(), provided_id.clone(), asset.clone(), direction, trigger_percentage)?;
+			let task: Task<T> = Task::<T> {
+				owner_id: who.clone(),
+				provided_id,
+				asset: asset.clone(),
+				direction,
+				trigger_percentage,
+			};
+			<Tasks<T>>::insert((asset, task_id), task);
+
+			Self::deposit_event(Event::TaskScheduled { who, task_id });
+			Ok(())
+		}
+
+		pub fn move_scheduled_tasks(
+			asset: Vec<u8>,
+			lower: u128,
+			higher: u128,
+			direction: u8,
+		) -> DispatchResult {
+			let mut existing_task_queue: Vec<(Vec<u8>, T::Hash)> = Self::get_task_queue();
+			// TODO: fix adjusted_higher to not peg to 20. Should move with the removal of 100 % increase.
+			let adjusted_higher = match higher > 20 {
+				true => 20,
+				false => higher,
+			};
+			for percentage in lower..adjusted_higher {
+				// TODO: pull all and cycle through in memory
+				if let Some(asset_tasks) = Self::get_scheduled_tasks((asset.clone(), direction.clone(), percentage.clone())) {
```

Comment:
this is a lot of reads. ideally i figure out how to pull into memory and cycle through and reinsert large parts of the storage map. but idk how to do that right now